Story/cite 204

PradnyaC11 opened the pull request:
# Guidelines for Pull Requests

If you haven't yet read our code review guidelines, please do so, You can find them [here](https://diging.atlassian.net/wiki/spaces/DIGING/pages/2256076801/Code+Review+Guidelines).

Please confirm the following by adding an x for each item (turn `[ ]` into `[x]`).

- [x] I have removed all code style changes that are not necessary (e.g. changing blanks across the whole file that don’t need to be changed, adding empty lines in parts other than your own code)
- [x] I am not making any changes to files that don’t have any effect (e.g. imports added that don’t need to be added)
- [x] I do not have any sysout statements in my code or commented out code that isn’t needed anymore
- [x] I am not reformatting any files in the wrong format or without cause. 
- [x] I am not changing file encoding or line endings to something else than UTF-8, LF
- [x] My pull request does not show an insane amount of files being changed although my ticket only requires a few files being changed
- [x] I have added Javadoc/documentation where appropriate
- [x] I have added test cases where appropriate
- [x] I have explained any part of my code/implementation decisions that is not be self-explanatory

## Please provide a brief description of your ticket 

... Put ticket description here and add link to ticket ...
Implementing an 'Add Reference' Button for All Groups in Citesphere with a Search Field Feature to Easily Add References to Group's Reference Field

https://diging.atlassian.net/browse/CITE-204

## Are there any other pull requests that this one depends on?
  
## Anything else the reviewer needs to know?

... describe here ...

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ReferenceController.java:
```diff
@@ -0,0 +1,35 @@
+            @RequestParam(value = "searchTerm", required = false) String searchTerm,
+            @RequestParam(defaultValue = "0", required = false, value = "page") int page,
+            @RequestParam(defaultValue = "20", required = false, value = "pageSize") int pageSize) {
+        return new ResponseEntity<ResultPage>(engine.search(searchTerm, zoteroGroupId, Math.max(page - 1, 0), 50), HttpStatus.OK);
```
what's 50? The pagesize? that should not be hard-coded

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/UpdateItemController.java:
```diff
@@ -0,0 +1,67 @@
+import edu.asu.diging.citesphere.user.IUser;
+
+@Controller
+public class UpdateItemController {
```
We have an `EditItemController`, this name is just confusing. It's a very specific update it does, isn't it? So the name should reflect that.

PradnyaC11 replied:
Changed name of class to AddReference Controller

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/UpdateItemController.java:
```diff
@@ -0,0 +1,67 @@
+    @Autowired
+    private ICitationManager citationManager;
+
+    @RequestMapping(value = "/auth/group/{zoteroGroupId}/items/{itemId}/update", method = RequestMethod.POST)
```
same thing, this url is too confusing.

PradnyaC11 replied:
changed the url from update to addReference

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/UpdateItemController.java:
```diff
@@ -0,0 +1,67 @@
+    private ICitationManager citationManager;
+
+    @RequestMapping(value = "/auth/group/{zoteroGroupId}/items/{itemId}/update", method = RequestMethod.POST)
+    public ResponseEntity<?> updateReference(Authentication authentication,
```
bad method name: the method does not update the reference

PradnyaC11 replied:
Changed updateReference to addReference

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/UpdateItemController.java:
```diff
@@ -0,0 +1,67 @@
+            ICitation citation = citationManager.getCitation((IUser) authentication.getPrincipal(), zoteroGroupId,
+                    itemId);
+            citation = citationManager.updateCitationReference(citation, referenceCitationKey, reference);
+            citationManager.updateCitation((IUser) authentication.getPrincipal(), zoteroGroupId, citation);
```
what do `updateCitationReference` and `updateCitation` do? why do they both need to be called? if they do things that should always be done together, then this is business logic and should not be in the controller.

PradnyaC11 replied:
updateCitation method is used at multiple places so, didn't change it. The method saves changed citation data to zetero.  updateCitationReference name is changed to addCitationToReferences, which adds found citation to references of the editing citation.

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/UpdateItemController.java:
```diff
@@ -0,0 +1,67 @@
+            return new ResponseEntity<>("{\"error\": \"" + e.getMessage() + "\"}", HttpStatus.INTERNAL_SERVER_ERROR);
+        } catch (CitationIsOutdatedException e) {
+            logger.error("Citation is outdated.", e);
+            return new ResponseEntity<>("{\"error\": \"Item " + itemId + " is outdated.\"}", HttpStatus.NOT_FOUND);
```
this should probably be a 409 Conflict

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html:
```diff
@@ -1,31 +1,38 @@
-			var identType = $(clicked).attr('data-identifier-type');
-			if (identType != undefined && identType != "") {
-				identifier = identifier + " (" + identType + ")";
+			citationID = $(clicked).attr('data-citation-id');
```
this creates a global variable, doesn't it? That should not be needed.

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html:
```diff
@@ -1,31 +1,38 @@
+			citationID = $(clicked).attr('data-citation-id');
+			if(citationID != undefined && citationID != "") {
+				zoteroGroupId = $(clicked).attr('data-zotero-group-id') 
+				window.location.href = "../items/" + citationID;
```
no, needs to be a properly generated url

PradnyaC11 commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:
```diff
@@ -489,9 +501,25 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
     }
+
+    @Override
+    public ICitation updateCitationReference(ICitation citation, String referenceCitationKey, String reference) {
```
Name of method changed to addCitationToReferences

PradnyaC11 commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:
```diff
@@ -489,9 +501,25 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
+        if (references == null) {
+            references = new HashSet<>();
+        }
+        if (!citation.getKey().equals(referenceCitationKey)) {
```
It checks if the citation is added to its own reference list.  Added an exception for this condition.

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:
```diff
@@ -494,4 +504,22 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
+            references = new HashSet<>();
+        }
+        if (!citation.getKey().equals(referenceCitationKey)) {
+            IReference iReference = new Reference();
```
why is this called `iReference`?

PradnyaC11 replied:
IReference is interface and it's implementation is Reference for reference of the citation. It was already present in citesphere-model.

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:
```diff
@@ -84,6 +90,9 @@ public class CitationManager implements ICitationManager {
     private IAsyncCitationProcessor asyncCitationProcessor;
 
+    @Autowired
+    private CitationRepository citationRepository;
```
this should not be used here directly but the `citationStore`

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/AddReferenceController.java:
```diff
@@ -0,0 +1,71 @@
+            ICitation citation = citationManager.getCitation((IUser) authentication.getPrincipal(), zoteroGroupId,
+                    itemId);
+            citation = citationManager.addCitationToReferences(citation, referenceCitationKey, reference);
+            citationManager.updateCitation((IUser) authentication.getPrincipal(), zoteroGroupId, citation);
```
why does the citation needs to be updated? has anything changed?

PradnyaC11 replied:
updateCitation method is called here to update the newly added reference in zetero too. Removing this does not add the reference to zetero's notes.

jdamerow replied:
`addCitationToReferences` should save the citation, there shouldn't be an extra call needed (especially not form the controller).

jdamerow replied:
still still

PradnyaC11 replied:
Moved it to addCitationReference.

jdamerow commented on citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/CitationManagerTest.java:
please read and follow our testing guidelines/naming conventions

PradnyaC11 commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ReferenceController.java:
```diff
@@ -0,0 +1,35 @@
+
+@Controller
+@PropertySource("classpath:/config.properties")
+public class ReferenceController {
```
Changed name of controller to SearchReferencesController, method name to - searchReferences  and URL to - /auth/group/{zoteroGroupId}/references?searchTerm={searchTerm}

jdamerow replied:
But it searches citations not references.

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:
```diff
@@ -494,4 +500,22 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
+    public ICitation addCitationToReferences(ICitation citation, String referenceCitationKey, String reference) throws SelfCitationException {
+        Set<IReference> references = citation.getReferences();
+        if (references == null) {
+            references = new HashSet<>();
```
if this is the case, then `references` will never be added to the citation.

PradnyaC11 replied:
This condition is added for the citations that don't have any references already added to them. In that case, there's a probability the references might be set to null. It occurs when a person create the citation, instead of importing it.

jdamerow replied:
yes, I understand that, but right now the references of the citation object will still be null (you never set the references on the citation, just the local variable)

PradnyaC11 replied:
Removing the condition gave me null pointer exception error when I tried to add reference to a dummy citation that I created in zetero. It does not give error when the new dummy citation is created in citesphere, because then it's initialized to empty array.

jdamerow replied:
What I'm saying is that this code:
```
Set<IReference> references = citation.getReferences();
if (references == null) {
            references = new HashSet<>();
}
...
 references.add(newReference);
```
will create a list of references that you can add to. However, if `citation.getReferences()` returned null before the if, it will still return null after the if. The following code wont' throw a nullpointer, but the changes you make to `references` won't be stored.

PradnyaC11 replied:
Got the logical error here. I have added a line to set references of citation so that the case you are saying won't happen.

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/ItemController.java:
```diff
@@ -38,6 +38,7 @@ public String getItem(Authentication authentication, Model model, @PathVariable(
             @RequestParam(required = false, defaultValue = "", value = "conceptIds") String[] conceptIds) throws GroupDoesNotExistException, CannotFindCitationException, ZoteroHttpStatusException {
         ICitation citation = citationManager.getCitation((IUser)authentication.getPrincipal(), zoteroGroupId, itemId);
         model.addAttribute("zoteroGroupId", zoteroGroupId);
+        model.addAttribute("itemId", itemId);
```
why is this needed? `itemId` is the id of the citation, isn't it? And the citation object is already in the model.

PradnyaC11 replied:
Removed itemId.

jdamerow commented on citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/CitationManagerTest.java:
```diff
@@ -69,7 +76,13 @@ public class CitationManagerTest {
+    private final String CITATION_KEY = "citationKey";
+    private ICitation updatedCitation;
+    private  Set<IReference> references;
+    private IReference iReference;
```
`iReference` is a bad variable name. What would the `i` stand for?

diging-jenkins commented:
Jenkins successfully deployed Citesphere to be reviewed!

jdamerow commented:
- layout is off
![image](https://github.com/diging/citesphere/assets/8881141/29c0a4e3-4bf2-4c9c-8cc8-e3c4fcff6529)

- The author for the references string seems to be taken randomly
![image](https://github.com/diging/citesphere/assets/8881141/1e8e0829-6a01-4967-8f1e-138d0f2de991)
The next time it's "Haines R, 2019, What Makes Research Software Sustainable? An Interview Study With Research Software Engineers"
If there are multiple authors, it should be "First author et al."
- then clicking "Add", it takes a long time where the browser is not reacting. If the user clicks again, it adds the reference twice. There needs to be a loading indicator and the user should not be able to click the button again.
- And I can add the same item as reference

PradnyaC11 commented:
> * layout is off
>   ![image](https://private-user-images.githubusercontent.com/8881141/342400001-29c0a4e3-4bf2-4c9c-8cc8-e3c4fcff6529.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkzNTIzNDksIm5iZiI6MTcxOTM1MjA0OSwicGF0aCI6Ii84ODgxMTQxLzM0MjQwMDAwMS0yOWMwYTRlMy00YmYyLTRjOWMtOGNjOC1lM2M0ZmNmZjY1MjkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYyNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MjVUMjE0NzI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjlhMTg1MjgzZjcwMzkxYWIwMWIwYWZjNGUxNGQzN2VhYzBlZjlhOWZkYTdkNmY1OTRiMDI2YTg2NTBlNjhhMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.5jHaiHvsKpKi99f3hj4RvLTWF6Jj5uN07FhHD_52zEM)
> * The author for the references string seems to be taken randomly
>   ![image](https://private-user-images.githubusercontent.com/8881141/342400623-1e8e0829-6a01-4967-8f1e-138d0f2de991.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkzNTIzNDksIm5iZiI6MTcxOTM1MjA0OSwicGF0aCI6Ii84ODgxMTQxLzM0MjQwMDYyMy0xZThlMDgyOS02YTAxLTQ5NjctOGYxZS0xMzhkMGYyZGU5OTEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYyNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MjVUMjE0NzI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Zjg3MmQ5OWRjYjdiZWJmYWMyMWJkOGEzMzVhNjIyZjkyOWU4ZmY2NTYyMGRiYzM1ZDY1ODA2MDk1OTFlMjBlYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.KPHQ1OG1ZINbeUONjqM5SHWYTUqovWY9nEyfD2yMZ-0)
>   The next time it's "Haines R, 2019, What Makes Research Software Sustainable? An Interview Study With Research Software Engineers"
>   If there are multiple authors, it should be "First author et al."
> * then clicking "Add", it takes a long time where the browser is not reacting. If the user clicks again, it adds the reference twice. There needs to be a loading indicator and the user should not be able to click the button again.
> * And I can add the same item as reference

1. Updated the layout. 
2. For random author creation problem, I have sort the list by firstname. 
3. Added the loaded spinner. 
4. I tested adding same item as reference, It threw exception as expected. I tested it on item - "Self-Organization in Pattern Formation".

diging-jenkins commented:
Jenkins successfully deployed Citesphere to be reviewed!

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html:
```diff
@@ -104,7 +115,170 @@
+						}
+						reference += ", "
+					}
+	                reference += (citationItem["dateFreetext"] != null && citationItem["dateFreetext"] != "") ? citationItem["dateFreetext"] + ", " : "";
```
all of this should not be done in javascript. reuse the code that generates the titles on the item pages here.

PradnyaC11 replied:
Created a template in the same file, as short form of citation is not being used at any other place and used it for title and at this place.

diging-jenkins commented:
Jenkins successfully deployed Citesphere to be reviewed!

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:
```diff
@@ -494,4 +498,28 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
+            newReference.setCitationId(referenceCitationKey);
+            newReference.setReferenceString(reference);
+            references.add(newReference);
+            citation.setReferences(references);
```
this should happen in line 510

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/AddReferenceController.java:
```diff
@@ -0,0 +1,70 @@
+    @Autowired
+    private ICitationManager citationManager;
+
+    @RequestMapping(value = "/auth/group/{zoteroGroupId}/items/{itemId}/addReference", method = RequestMethod.POST)
```
should be `.../references/add`

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/SearchCitationsController.java:
```diff
@@ -0,0 +1,31 @@
+    private SearchEngine engine;
+
+    @RequestMapping(value = { "/auth/group/{zoteroGroupId}/references" })
+    public ResponseEntity<ResultPage> searchReferences(Authentication authentication, @PathVariable String zoteroGroupId,
```
similar to the class name, this should not be called 'searchReferences' since it's not searching references, and the url should similarly be reflecting that it searches citations. Right now the url suggests it returns references.

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Build failed. Maybe you broke the test cases?

jdamerow commented:
compilation errors in tests

diging-jenkins commented:
Jenkins successfully deployed Citesphere to be reviewed!

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html:
```diff
@@ -104,7 +150,143 @@
+	            exit: 'animated fadeOutRight'
+	        },
+	    });
+	}
```
there are a couple of odd indentation issues

jdamerow replied:
I pointed out a few but there are a few more

jdamerow commented:
Make it so, Jenkins.

jdamerow commented:
- The search results box that shows the reference search results needs to show author and year, not just title.
- Also the box needs to say "Reference Search Result**s**"
- And the column with the add button should have a heading "Add as reference"
- The references input field in the "Add references" modal should be deleted; instead the modal should close when an add button is clicked.

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Jenkins successfully deployed Citesphere to be reviewed!

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Jenkins successfully deployed Citesphere to be reviewed!

jdamerow commented:
When clicking on a reference after it has been added, the pop up is all empty. it needs to pull the information from the citation.

jdamerow commented:
and there is a blank that shouldn't be there:
![image](https://github.com/user-attachments/assets/65d5c6ea-bf0c-4666-8333-f635140d30e2)

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:
```diff
@@ -494,4 +496,28 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
+            newReference.setCitationId(referenceCitationKey);
+            newReference.setReferenceString(reference);
+            references.add(newReference);
+        } else {
```
let's turn this if/else around so that we loose one indented block, so
```
if keys are equal:
      throw exception
code if they are not equal goes here
```

diging-jenkins commented:
Jenkins successfully deployed Citesphere to be reviewed!

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:
```diff
@@ -494,4 +496,28 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
+            references = new HashSet<>();
+            citation.setReferences(references);
+        }
+        if (citation.getKey().equals(referenceCitationKey)) {
```
this should happen first thing. Nothing of the code above needs to be executed if an exception is thrown here.

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/CitationManager.java:
```diff
@@ -494,4 +496,28 @@ public CitationPage getPrevAndNextCitation(IUser user, String groupId, String co
+        Set<IReference> references = citation.getReferences();
+        if (references == null) {
+            references = new HashSet<>();
+            citation.setReferences(references);
```
merge with line 507

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html:
```diff
@@ -104,7 +134,141 @@
+			complete: function() {
+				$("#addReferenceSpinner").hide();
+				$("#referenceModal").modal('hide');
+			}
```
indentation

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html:
```diff
@@ -104,7 +134,141 @@
+	    
+	    $("#addReferences").click(function() {
+	        $("#citationReference").val('');
+			$("#itemReference")[0].innerHTML = '';
```
indentation

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html:
```diff
@@ -104,7 +134,141 @@
+	        $("#addReferenceSpinner").show();
+	    	document.getElementById('addReferenceSpinner').style.display = 'inline-block';
+
+			var zoteroGroupId = $("#addReferences").data('zotero-group-id');
```
indent

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Jenkins successfully deployed Citesphere to be reviewed!

jdamerow commented:
- When there are no search results, there should be a message that says so.
- search should also happen if you hit enter after entring a term

diging-jenkins commented:
Build failed. Maybe you broke the test cases?

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Jenkins successfully deployed Citesphere to be reviewed!

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/group/item.html:
```diff
@@ -104,7 +134,156 @@
+	            	$("#citationSearchTable").empty();
+	            	var content = '<div class="alert warning">'
+	            	content += '<strong>Warning!</strong>'
+	            	content += ' No citations found for the search query. Try other search query.'
```
The message should just be: "No citations were found for the entered search term."
(No "warning" no "try other search query")

jdamerow reviewed (changes requested):
After that should be good.

jdamerow commented:
Also there is a bug, if I search for soemthing that yields results, it shows results. I click close without selecting something. I search for something that doesn't exist, it shows me no results. I click close. I search for something that should yield results, but now it shows me no results.

diging-jenkins commented:
Jenkins successfully deployed Citesphere to be reviewed!

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Jenkins successfully deployed Citesphere to be reviewed!

jdamerow commented:
Another bug:
- I search for a term, it shows results
- I select one
- I don't click "Add reference" but search again
- I don't select anything but just click "close"
- the reference I selected before is still shown, but the 'add reference' button is now disabled
=> I should be able to select the reference I selected first

diging-jenkins commented:
Jenkins successfully deployed Citesphere to be reviewed!

diging-jenkins commented:
Jenkins successfully deployed Citesphere to be reviewed!
